libcxx/src/locale.cpp

```diff
@@ -538,16 +543,31 @@ locale::__imp::use_facet(long id) const
 
 // locale
 
+std::atomic<locale::__imp*> locale::__imp::classic_;
+
 const locale&
 locale::__imp::make_classic()
 {
     // only one thread can get in here and it only gets in once
     alignas(locale) static std::byte buf[sizeof(locale)];
     locale* c = reinterpret_cast<locale*>(&buf);
     c->__locale_ = &make<__imp>(1u);
+    classic_.store(c->__locale_, std::memory_order_relaxed);
```

Comment:
I only see `memory_order_relaxed` in the usage. How are the atomic reads and writes synchronized?

Reply:
Added a comment explaining why relaxed ordering is the right one.

Reply:
I agree we only care about the value of the pointer and not what information deferring the pointer gives. I agree this should happen
```
    // If a locale uses the classic imp, then this store happens
    // before acquire/release methods, and they must observe the
    // right value and omit reference counting.
```
However I don't see any guarantee this does happen. Per https://en.cppreference.com/w/cpp/atomic/memory_order
```
Relaxed operation: there are no synchronization or ordering constraints imposed on other reads or writes, only this operation's atomicity is guaranteed (see [Relaxed ordering](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering) below).
```
Specifically I see no guarantee that when Thread A creates the classic locale and thread B uses it the classic locale the value of `classic_` read in Thread B is the value stored in Thread A. Can you explain what I'm missing?

Reply:
The visibility of classic_ is provided by the dynamic local static initialization of c in this case:
```
static const locale& c = __imp::make_classic();
```
Any thread that called classic() and passed though this statement must have visibility over c and and any side effects of make_classic(). The store to classic_ is a side effect of make_classic().

I can't find the wording in the standard that says that completion of the dynamic initialization happens-before any uses of the variable, but it's definitely the intention and the code would be broken today otherwise (along with tons of other C++ code out there).

Think of this local static initialization as call_one:
```
32.5.6.2
3 Synchronization: For any given once_flag: all active executions occur in a total order; completion
of an active execution synchronizes with (6.9.2) the start of the next one in this total order; and the
returning execution synchronizes with the return from all passive executions.
```